src/aal/aal_sparc.h

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#if defined(__arch64__) // More reliable than __sparc64__
+#  define SNMALLOC_VA_BITS_64
+#else
+#  define SNMALLOC_VA_BITS_32
+#endif
+
+namespace snmalloc
+{
+  /**
+   * Sparc architecture abstraction layer.
+   */
+  class AAL_Sparc
+  {
+  public:
+    /**
+     * Bitmap of AalFeature flags
+     */
+    static constexpr uint64_t aal_features = StrictProvenance;
+
+    static constexpr enum AalName aal_name = Sparc;
+
+#ifdef SNMALLOC_VA_BITS_64
+    /**
+     * Even Ultra-Sparc I supports 8192 and onwards
+     */
+    static constexpr size_t smallest_page_size = 0x2000;
+#else
+    static constexpr size_t smallest_page_size = 0x1000;
+#endif
+
+    using address_t = uintptr_t;
+
+    /**
+     * On Sparc ideally pause instructions ought to be
+     * optimised per Sparc processor but here a version
+     * as least common denominator to avoid numerous ifdef,
+     * reading Conditions Code Register here
+     */
+    static inline void pause()
+    {
+      __asm__ volatile("rd %ccr, %g0 \n\trd %ccr, %g0 \n\trd %ccr, %g0");
+    }
+
+    static inline void prefetch(void* ptr)
+    {
+#ifdef SNMALLOC_VA_BITS_64
+      __asm__ volatile("prefetch [%0], 0" ::"r"(ptr));
+#else
+      (void)ptr;
+#endif
+    }
+
+    static inline uint64_t tick()
+    {
+      uint64_t tick;
+      __asm__ volatile(".byte 0x83, 0x41, 0x00, 0x00");
+      __asm__ volatile("mov %%g1, %0" : "=r"(tick));
```

Comment:
Why is this written disassembled and why is it two asm blocks?  `asm volatile` may produce side effects and may not be discarded, but the compiler *is* free to move things between the blocks and so there's no guarantee that the two are adjacent and that %g1 is not clobbered between the two.  

The LLVM disassembler can disassemble this, so I'd assume that both gas and the LLVM assembler can handle assembling it.

```suggestion
      __asm__ volatile("rd %asr4, %0" : "=r"(tick)");
```

Reply:
Right it looks simpler. In fact I stumbled across it in an old sparc book assembly but let s go with your version.